General refactoring

MBalszun opened the pull request:
Collection of fixes / improvements that afaik don't depend on c++11. 

Notes: 
- Includes https://github.com/mbeckersys/MavLogAnalyzer/pull/3 but doesn't depen
- The commits are independent, so you can pick and choose / squash / rearrange as you like

MBalszun committed:
Remove const cast mumbo jumbo in _addDataToPlot

- Top level const modifier has no effect on pointers when used in
dynamic cast, so remove them

- Original version first casts a pointer to non-const to
  a pointer to const and then passes that to a function taking
  pointer to non-const.
  -> Aside from being unnecessary (just stick with pointer to non-const),
     any self-respecting compiler should produce an error
     on the last operation ...

MBalszun committed:
Small modifications to improve compatibility and reduce type conversions

MBalszun committed:
Further improve parse performance by avoiding repeatedly computing parameter path for each data point

MBalszun commented:
OK, I rebased the PR and (hopefully) removed all c++11 parts (the result is now much smaller than the original PR)

mbeckersys commented on src/mainwindow.h:
```diff
@@ -116,7 +116,7 @@ private slots:
 
 private:
-    void _addDataToPlot(TreeItem * const item);
+    void _addDataToPlot(TreeItem * item);
```
That should stay `TreeItem * const item` otherwise signature/definition mismatch (some IDEs don't match their tags anymore).

MBalszun replied:
OK

mbeckersys commented on src/mavlinkscenario.cpp:
```diff
@@ -285,34 +285,42 @@ bool MavlinkScenario::add_onboard_message(const OnboardData &msg) {
     for (OnboardData::booldata_t::const_iterator dit = msg.get_booldata().begin(); dit != msg.get_booldata().end(); ++dit) {
         if (dit->first == "t") continue;
-        string fullname = "onboard log/" + msg.get_message_name() + "/" + dit->first;
+        fullname.resize(stem_len);
```
Is `resize` followed by `append` any faster? Otherwise I would prefer to just use `append`, to keep the code shorter.

MBalszun replied:
Ok, that probably needs more explanation in the code:
- there is now only a single local `std::string` variable that is passed to all calls of `track_generic_timeseries` (before, multiple temporary strings were created for every single call, resulting in a lot of dynamic memory allocations).
- The main part, consisting of `"onboard log/" + msg.get_message_name() + "/"` stayes the same for all of those calls, but the last part (`dit->first`) has to be changed for every call
- The resize function SHORTENS the string such that only the constant (stem) part remains (the internal buffer/capacity of course stays the same size)
- The call to append then only copies a few new characters into the internal buffer space.

Instead of `resize` one could probably call `erase` to clarify the purpose.

mbeckersys commented on src/data_event.h:
```diff
@@ -105,7 +105,7 @@ class DataEvent : public DataTimed {
     unsigned long get_epoch_dataend() const {
         if (_elems_time.empty()) { return get_epoch_datastart(); }
-        return ((unsigned long) _elems_time.back()*1E6) + get_epoch_datastart();
+        return (unsigned long)( _elems_time.back()* 1E6) + get_epoch_datastart();
```
good catch

MBalszun committed:
Address review comments by martin becker

- Put constness of pointer back into function signature
- Make changes in MavlinkScenario::add_onboard_message hopefully more understandable
